clients/unshaded/src/main/java/tachyon/client/block/BufferedBlockInStream.java

```diff
@@ -0,0 +1,244 @@
+    mPos = pos;
+  }
+
+  @Override
+  public long skip(long n) throws IOException {
+    checkIfClosed();
+    if (n <= 0) {
+      return 0;
+    }
+
+    long toSkip = Math.min(remaining(), n);
+    mBufferPos = INVALID_BUFFER_POS;
+    mPos += toSkip;
+    return toSkip;
+  }
+
+  /**
+   * Reads from the data source into the buffer. The buffer should be at position 0 and have len
+   * valid bytes available after this method is called. This method should not modify mBufferPos,
+   * mPos, or increment any metrics.
+   *
+   * @param len length of data to fill in the buffer, must always be <= buffer size
+   * @throws IOException if the read failed to buffer the requested number of bytes
+   */
+  protected abstract void bufferedRead(int len) throws IOException;
+
+  /**
+   * Directly reads data to the given byte array. The data will not go through the internal buffer.
+   * This method should not modify mPos or update any metrics collection for bytes read.
+   *
+   * @param b the byte array to write the data to
+   * @param off the offset in the array to write to
+   * @param len the length of data to write into the array must always be valid within the block
+   * @return the number of bytes successfully read
+   * @throws IOException if an error occurs reading the data
+   */
+  protected abstract int directRead(byte[] b, int off, int len) throws IOException;
+
+  /**
+   * Increments the number of bytes read metric. Inheriting classes should implement this to
+   * increment the correct metric.
+   *
+   * @param bytes number of bytes to record as read
+   */
+  protected abstract void incrementBytesReadMetric(int bytes);
+
+  /**
+   * Initializes the internal buffer based on the user's specified size. Any reads above half
+   * this size will not be buffered.
+   *
+   * @return a heap buffer of user configured size
+   */
+  private ByteBuffer allocateBuffer() {
+    TachyonConf conf = ClientContext.getConf();
+    return ByteBuffer.allocate((int) conf.getBytes(Constants.USER_REMOTE_READ_BUFFER_SIZE_BYTE));
+  }
+
+  /**
+   * Convenience method to ensure the stream is not closed.
+   */
+  private void checkIfClosed() {
+    Preconditions.checkState(!mClosed, "Cannot do operations on a closed BlockInStream");
+  }
+
+  /**
+   * @return how many bytes are still valid in the buffer.
+   */
+  protected long remainingInBuffer() {
+    if (mBufferPos == INVALID_BUFFER_POS) {
+      return 0;
+    }
+
+    return mBuffer.remaining();
+  }
+
+  /**
+   * Updates the buffer so it is ready to be read from. After calling this method, the buffer will
+   * be positioned at 0 and mBufferPos will be equal to mPos. Inheriting classes should implement
+   * bufferedRead(int) for their read specific logic.
+   *
+   * @throws IOException if an error occurs reading the data
+   */
+  private void updateBuffer() throws IOException {
+    int toRead = (int) Math.min(mBuffer.limit(), remaining());
+    bufferedRead(toRead);
+    mBufferPos = mPos;
```

Comment:
After `bufferedRead`, should we check the resulting buffer size is really `toRead` bytes?